zfit/models/kde.py

```diff
@@ -0,0 +1,134 @@
+#  Copyright (c) 2020 zfit
+from typing import Union
+
+import numpy as np
+import tensorflow as tf
+import tensorflow_probability as tfp
+from tensorflow_probability.python import distributions as tfd
+
+from .dist_tfp import WrapDistribution
+from .. import z, ztypes
+from ..core.interfaces import ZfitData
+from ..util import ztyping
+from ..util.exception import OverdefinedError, ShapeIncompatibleError
+
+
+def bandwidth_rule_of_thumb(data, factor=0.9):
+    return tf.math.reduce_std(data) * tf.cast(tf.shape(data)[0], ztypes.float) ** (-1 / 5.) * factor
+
+
+def bandwidth_silverman(data):
+    return bandwidth_rule_of_thumb(data=data, factor=0.9)
+
+
+def bandwidth_scott(data):
+    return bandwidth_rule_of_thumb(data=data, factor=1.059)
+
+
+def bandwidth_adaptiveV1(data, func):
+    from .. import run
+    run.assert_executing_eagerly()
+    data = z.convert_to_tensor(data)
+    bandwidth = z.sqrt(tf.math.reduce_std(data) / func(data))
+    bandwidth *= tf.cast(tf.shape(data)[0], ztypes.float) ** (-1 / 5.) * 1.059
```

Comment:
Just a curiosity what kind of adaptive bandwidth did you implement here?. I see many things in the literature but the most common on is $$h(x) = \frac{h_0}{pdf(x)^2 \times G}  $$ where $$pdf$$ is the first estimate of density as you did and $$G$$ is the geometric mean of the data.

Reply:
I took eq. 6 from [Kyles Kernel Density in HEP](https://reader.elsevier.com/reader/sd/pii/S0010465500002435?token=BCE1BEFF1924EF76532CCF9D2D3F0274B4C41E49DB20B0CBFEA2A32AE5FD51826548F615DCCEA29D5B4AA65EF524DFD0) (modulo using also the IQR, should put that there). However, that square root and why "it clearly follows" is (intuitively) not fully clear to me.

Where did you find the geometric mean? I would not quite understand why we would need that for the bandwidth (the spread)?

Reply:
This is how it is done in [Stata](https://www.stata.com/meeting/9uk/akdensity.pdf) but this was also described in a Silverman's book (searching for a link).

Reply:
> I took eq. 6 from [Kyles Kernel Density in HEP](https://reader.elsevier.com/reader/sd/pii/S0010465500002435?token=BCE1BEFF1924EF76532CCF9D2D3F0274B4C41E49DB20B0CBFEA2A32AE5FD51826548F615DCCEA29D5B4AA65EF524DFD0) (modulo using also the IQR, should put that there). However, that square root and why "it clearly follows" is (intuitively) not fully clear to me.
> 
> Where did you find the geometric mean? I would not quite understand why we would need that for the bandwidth (the spread)?
 
You mean that `h* ∝ σ **2` ?. This I succeeded to convince me about that since `σ ∝ n**0.5`, but still I don't get how Eq. 6 is obtained.

Reply:
Anyway as long as new bandwidth scales with the inverse of the pilot pdf scare it's fine, then it's juste fine tuning. The reason for the geometric mean is explained in Silverman .
<img width="704" alt="Capture d’écran 2020-06-11 à 12 09 20" src="https://user-images.githubusercontent.com/24250309/84373097-6393b200-abdc-11ea-8a68-885e3bcb226b.png">


<img width="723" alt="Capture d’écran 2020-06-11 à 12 08 59" src="https://user-images.githubusercontent.com/24250309/84373063-57a7f000-abdc-11ea-8919-b33cb463af9e.png">

Reply:
I see, G refers to the geometric mean of the pilot estimate, e.g. of f_0(x_i), not of the data points (so kinda the average density response of our first estimate)

Reply:
So here two things are used, lambda and h, currently we place them together. It could surely be cleaner though to have both, I agree

Reply:
Ah right I said the geometric mean on the data, but it's indeed on the pilot estimate.  Yeah having both h and λ = 1 by default for fixed kernel bandwidth, and then λ = something / √𝓅𝒹𝒻 for adaptive.

Reply:
While I agree on the nice splitting into h and lambda, it seems a little bit more complicated to generalize properly with parameters, as there are currently only single valued parameters and how to handle it. My idea is to go with the simpler layout (currently) of having just a bandwidth that can be specified. While this is somewhat limited, we can get some experience with it and then better generalize (also to ndims). Sounds reasonable?

Reply:
Yes it's reasonable, improvements can come later.